File: leetCode/fukun/lc_064.js
const minPathSum = function (dp) {
    for (let i = 0; i < dp.length; i++) {
        for (let j = 0; j < dp[0].length; j++) {
            if (i === 0) {
                if (j !== 0) {
                    dp[i][j] += dp[i][j - 1];
                }
            } else {
                if (j === 0) {
                    dp[i][j] += dp[i - 1][j];
                } else {
                    dp[i][j] = Math.min(dp[i - 1][j], dp[i][j - 1]) + dp[i][j];
                }
            }
        }
    }
    return dp.pop().pop();
};
console.log(minPathSum([
    [1, 3, 1],
    [1, 5, 1],
    [4, 2, 1]
]));

File: leetCode/lwc/lc_064.java
package solution2;

/**
 * @author lwc
 * @date 2019/10/12 10:00
 */
public class U64 {
    public int minPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int i = 1; int j = 1;
        while(j < n) grid[0][j] = grid[0][j-1] + grid[0][j++];
        while(i < m) grid[i][0] = grid[i-1][0] + grid[i++][0];
        for(i = 1; i < m; i++){
            for(j = 1; j < n; j++){
                grid[i][j] = Math.min(grid[i][j-1],grid[i-1][j]) + grid[i][j];
            }
        }
        return grid[m-1][n-1];
    }
}
